WIP refactor:improve checkpoint and ensure gc to improve disk space

gauravsaini opened the pull request:
Still work in progress

<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../CONTRIBUTING.md).
-->

### Related GitHub Issue

<!-- Every PR MUST be linked to an approved issue. -->

Closes: # <!-- Replace with the issue number, e.g., Closes: #123 -->

### Description

<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

### Type of Change

<!-- Mark all applicable boxes with an 'x'. -->

- [ ] 🐛 **Bug Fix**: Non-breaking change that fixes an issue.
- [ ] ✨ **New Feature**: Non-breaking change that adds functionality.
- [ ] 💥 **Breaking Change**: Fix or feature that would cause existing functionality to not work as expected.
- [ ] ♻️ **Refactor**: Code change that neither fixes a bug nor adds a feature.
- [ ] 💅 **Style**: Changes that do not affect the meaning of the code (white-space, formatting, etc.).
- [ ] 📚 **Documentation**: Updates to documentation files.
- [ ] ⚙️ **Build/CI**: Changes to the build process or CI configuration.
- [ ] 🧹 **Chore**: Other changes that don't modify `src` or test files.

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [ ] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [ ] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [ ] **Self-Review**: I have performed a thorough self-review of my code.
- [ ] **Code Quality**:
    - [ ] My code adheres to the project's style guidelines.
    - [ ] There are no new linting errors or warnings (`npm run lint`).
    - [ ] All debug code (e.g., `console.log`) has been removed.
- [ ] **Testing**:
    - [ ] New and/or updated tests have been added to cover my changes.
    - [ ] All tests pass locally (`npm test`).
    - [ ] The application builds successfully with my changes.
- [ ] **Branch Hygiene**: My branch is up-to-date (rebased) with the `main` branch.
- [ ] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [ ] **Changeset**: A changeset has been created using `npm run changeset` if this PR includes user-facing changes or dependency updates.
- [ ] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](../CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->

hannesrudolph commented:
@gauravsaini would you be able to  convert the PR to draft and generate and link to an issue please?

gauravsaini commented:
> @gauravsaini would you be able to convert the PR to draft and generate and link to an issue please?
Hey, I've added unit tests to make sure that it works fine . I've also added possible issues it might fix. 
Ellipses dev bot gives a very good summary of the changes

hannesrudolph commented:
@cte would you be able to review this before it gets outdated? Looks promising.

daniel-lxs commented:
Hey @gauravsaini, I like this idea, I noticed a lot of comments like `// --- CHANGE START:` and `// --- CHANGE END:`, I think those are not necessary and can be removed.

daniel-lxs commented on src/services/checkpoints/ShadowCheckpointService.ts:
```diff
@@ -131,74 +161,140 @@ export abstract class ShadowCheckpointService extends EventEmitter {
+						// Copied
+						filesToAdd.push(filePath) // Add the new path
+					}
+					// Other statuses like 'U' (unmerged) might need specific handling if relevant
```
Hey @gauravsaini, quick thought on the new `stageAll` logic: how does it handle files that were part of a merge conflict and have been resolved in the working directory but not yet `git add`ed? The `if/else` chain doesn't seem to explicitly cover 'U' (unmerged) statuses. Just wondering if this might mean resolved conflicts aren't always included in checkpoints?

Also could this be changed to a `switch` statement?

KJ7LNW commented:
Question: Can `.git/` for checkpoints be moved to `.roo/checkpoints/.git` ?

I really like the idea of one out of band git tree per workspace, (eg .roo/.git) because then I could do something like `git remote add roo .roo/checkpoints`) and work with task branches accordingly. branch names like `roo/task/<TASK_ID>` could be renamed by users to be meaningful or AI could come up with a reasonable branch mean for each task something like `roo/task/<TASK_ID>-human-readable-name` . 

AI profiles could be given instructions to work with that repository to see and integrate into the existing main repository, mostly for cherry picks because it is only a partial file tree, but maybe other options as well.  Here are some other really cool ideas (o4-mini).

I especially like the feature that provide to ask the AI things like `search for the task that does X and integrate the change Y into this current task` --- I can not tell you how many times I have wanted to find something in my ancient task history, but it is seriously difficult because there is so much and the search tool really is not sufficient.

> By giving Roo a small set of Git primitives over `.roo/checkpoints.git` plus these inspection commands, you get full “time-travel” and merge strategies for just the files Roo touched—*and* you can ask the AI to audit, summarize, or triage any task before it ever lands in your main codebase.
> Here’s a much shorter sketch of how you’d teach Roo (and your VS Code extension) to work with a private “.roo/.git” alongside your main repo—and especially how to get it to *inspect* tasks before you ever merge anything:
> 
> 1. Core “Roo Git” Verbs  
>    • list-tasks  
>      – List all `roo/task/*` branches (with status: in-progress, done, abandoned).  
>    • diff-task  
>      – `git diff main…roo/task/123` (or any two checkpoints on that branch).  
>    • cherry-pick-task  
>      – Apply one or all commits from `roo/task/123` into your current work branch.  
>    • rebase-task  
>      – Rebase the task branch onto the latest `main`.  
>    • sandbox-task  
>      – Create a throw-away worktree: `git worktree add .roo/sandbox/123 roo/task/123` for isolated testing.  
>    • prune-tasks  
>      – Delete or archive stale/merged/abandoned task branches automatically.
> 
> 2. AI Inspection Hooks  
>    Teach Roo to *query* and *summarize* any task branch before you merge or prune it:
> 
>    a. change-summary  
>      – “What’s the high-level purpose of task 123?”  
>      – AI parses the full diff and emits a bullet-list: feature added, bug fixed, files/modules touched.  
> 
>    b. impact-analysis  
>      – “Which public APIs, config files, or documentation does task 123 affect?”  
>      – AI reads filenames/AST to highlight potential ripple effects.  
> 
>    c. conflict-risk  
>      – “Where might merging task 123 into `main` conflict?”  
>      – AI scans overlapping hunks or related modules and flags hotspots.  
> 
>    d. quality-audit  
>      – “Run lint/tests on sandbox 123 and summarize failures or coverage gaps.”  
>      – AI invokes your CI locally, collates errors, suggests fixes.  
> 
>    e. security/perf review  
>      – “Any new TODOs, insecure patterns, or performance regressions in task 123?”  
>      – AI greps for e.g. raw SQL, disables asserts, expensive loops, etc., then reports.  
> 
> 3. Common Inspection Scenarios  
>    • Before merging a big refactor—get a natural-language summary + risk map so you know where to spot-check.  
>    • On an abandoned branch—ask “Why did I abandon task 234?” and AI reviews commits/comments to surface the blocker.  
>    • When parallel tasks touch similar modules—have AI compare two branches and advise which to cherry-pick first.  
>    • After an AI-generated feature—“Generate a PR description and list missing tests” so you can complete the review.  
>    • During cleanup—“Which abandoned tasks haven’t run tests in 30 days?” or “Which feature tasks never got finished?”  
>

daniel-lxs commented:
This looks good to me, the checkpoints work normally on existing tasks so the implementation is backwards compatible.

daniel-lxs commented:
Hey @gauravsaini, it seems like the flag `--path-walk` used by your implementation:
```typescript
await git.raw(["repack", "-a", "-d", "-f", "--path-walk", "--quiet"])
```

Isn't available on my git version `2.49.0`

Do I need an specific version for this to work?

daniel-lxs commented:
> maybe we could omit this for compatibility

Hey @ijin, it seems like the author of the PR is unavailable, if you'd like you can create a PR based on this one with the compatibility changes, let's see how that turns out.

Let me know if you have any questions!

ijin commented:
> > maybe we could omit this for compatibility
> 
> Hey @ijin, it seems like the author of the PR is unavailable, if you'd like you can create a PR based on this one with the compatibility changes, let's see how that turns out.
> 
> Let me know if you have any questions!

@daniel-lxs I took a stab at it!
https://github.com/RooCodeInc/Roo-Code/pull/4494
